### PR TITLE
Return an error in `service.NewConfigProvider` if there is no configLocations (v0.45)

### PR DESCRIPTION
**Pull request by noisersup**
**Description:** 
As was said in #4734, this commit:  

- changes `service.NewConfigProvider` to return an error when provided `configLocations` length is equals to 0,
- deprecates `service.MustNewConfigProvider` and makes it `panic` under the same condition as above,
- changes back all references in code from "Must" version to normal and handles error.

When the version 0.44 will come out I will undraft this PR.
Also I'm not sure is it possible to apply only third commit (the previous 2 are from my last PR) but I will remove them when my last PR will be merged.

**Link to tracking Issue:** #4730 

**Testing:** `make all`

### CODE REVIEW

File: service/config_provider.go
Comment (bogdandrutu) [inline service/config_provider.go:88]: If you want to go this path, you need to "deprecate" the current `NewConfigProvider` and in 2 weeks we change that to return error.

Same for the "NewDefault", need a must version and "deprecate"

Comment (noisersup) [reply service/config_provider.go:88]: I deprecated it [here](https://github.com/open-telemetry/opentelemetry-collector/pull/4734/commits/22eb74f37ccd026ae0d8090575658a7e0bb57844)

The current PR was supposed to be the second step of the three given in this [comment](https://github.com/open-telemetry/opentelemetry-collector/pull/4734/commits/22eb74f37ccd026ae0d8090575658a7e0bb57844). It will be undraft after releasing version with the previous PR merged.

Also, sorry for an oversight at `NewDefaultConfigProvider` I'm going to push the commit with the must version to the [previous PR](https://github.com/open-telemetry/opentelemetry-collector/pull/4734).

**Comment by bogdandrutu** [comment]
Plese rebase.

**Comment by noisersup** [comment]
Should I add entry to changelog?

In my last PR `Skip Changelog` flag was added.

**Comment by bogdandrutu** [comment]
@noisersup yes add a changelog entry, and rebase so we can merge this.

**Comment by noisersup** [comment]
done

**Comment by bogdandrutu** [comment]
@noisersup I am sorry, I missed this release, but I think that is good, because I think we can do even better now. What about we start using an Option pattern here, and keep only the locations (as required) field?

cc @codeboten @tigrannajaryan @jpkrohling 

```
func NewConfigProvider(locations []string, opts ...ConfigProviderOption) (ConfigProvider, error) {
}
```

**Comment by bogdandrutu** [comment]
@noisersup looks like we have an agreement here, do you mind implementing that change?

**Comment by noisersup** [comment]
@bogdandrutu Sure, I'll implement that.

**Comment by noisersup** [comment]
I hope I understood the requirements

File: service/config_provider.go
Comment (bogdandrutu) [inline service/config_provider.go:94]: Not sure if we want a map as argument or one of the next two options:

1.
```golang
func WithConfigMapProvider(scheme string, p configmapprovider.Provider) ConfigProviderOption {
	return func(opts *ConfigProviderOptions) {
		opts.configMapProviders[scheme] = p
	}
}
```

2.
```golang

// Change configmapprovider.Provider to have a `Scheme()` func on it.

func WithConfigMapProvider(p configmapprovider.Provider) ConfigProviderOption {
	return func(opts *ConfigProviderOptions) {
		opts.configMapProviders[p.Scheme()] = p
	}
}
```

File: service/config_provider.go
Comment (bogdandrutu) [inline service/config_provider.go:100]: The order of converters matters, the question I have is: do we ask the user to give a slice of them or one by one:

```golang

func WithCfgMapConverters(c config.MapConverterFunc) ConfigProviderOption {
	return func(opts *ConfigProviderOptions) {
		opts.cfgMapConverters = append(opts.cfgMapConverters, c)
	}
}

return NewConfigProvider(
	configLocations,
	WithCfgMapConverter(configmapprovider.NewOverwritePropertiesConverter(properties)),
	WithCfgMapConverter(configmapprovider.NewExpandConverter())
	)
```

**Review by bogdandrutu — COMMENTED** [review]
I documented couple of options. Probably we can extract issues out of these and move forward with the current PR which does improve lots of the things.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:100]: I think this is not a good idea, since if we define a default "converter" user cannot remove/change that.

File: service/config_provider.go
Comment (bogdandrutu) [inline service/config_provider.go:84]: this type is not needed. You can change the option to be: `type ConfigProviderOption func(opts *configProvider)`. External user cannot and should not be able to extend the `ConfigProviderOption` anyway.

File: service/config_provider.go
Comment (bogdandrutu) [inline service/config_provider.go:174]: This should become the default values for `NewConfigProvider`. So I would remove this and change NewConfigProvider to have these map providers (user will be able to override this with the `WithCfgMapProviders`) and these map converters (user will be able to override this with the `WithCfgMapConverters`).

**Review by bogdandrutu — APPROVED** [review]

File: service/config_provider.go
Comment (noisersup) [reply service/config_provider.go]: I wanted to implement both approaches in seperate commits but I have an issue with the first.

Due to the fact that 2 of the functions (`service.MustNewConfigProvider` and `service.TestConfigProvider_Errors`) that use `WithConfigMapProviders` are using `map[string]configmapprovider.Provider` approach I need to translate these maps to `[]ConfigProviderOption` and after that append this array to other Options and pass them to `NewConfigProvider` function.

example:
```go
// Convert map[string]configmapprovider.Provider to []ConfigProviderOption

configMapProvidersOpts := []ConfigProviderOption{}
for scheme, provider := range configMapProviders {
configMapProvidersOpts = append(configMapProvidersOpts, WithConfigMapProvider(scheme, provider))
}

configProvider, err := NewConfigProvider(locations, append(configMapProvidersOpts, WithCfgMapConverters(cfgMapConverters), WithConfigUnmarshaler(configUnmarshaler))...)
```

should I create a function that converts config maps to `[]ConfigProviderOption` and appends them to provided options 
to avoid reused code or it's not that big deal?

```go
func AppendConfigMapProvider(
	opts []ConfigProviderOption,
	configMapProviders map[string]configmapprovider.Provider) []ConfigProviderOption {
	// Convert map[string]configmapprovider.Provider to []ConfigProviderOption
	for scheme, provider := range configMapProviders {
		opts = append(opts, WithConfigMapProvider(scheme, provider))
	}
	return opts
}

```

File: service/config_provider.go
Comment (Aneurysm9) [inline service/config_provider.go:92]:
```suggestion
func WithConfigMapConverters(c ...config.MapConverterFunc) ConfigProviderOption {
	return func(opts *configProvider) {
		opts.cfgMapConverters = append(opts.cfgMapConverters, c...)
	}
}
```

I think I would prefer to see these append to the config, allowing for multiple invocations of the options, possibly provided by separate packages.  The invocation of `NewConfigProvider()` ultimately controls the order in which they're applied.  A small change would be required there to set the default values after applying options if the list is empty, rather than setting it first and allowing it to be overwritten.

The same applies to `WithConfigMapProviders()`.  There order is not important, so multiple maps can be merged.

**Comment by Aneurysm9** [comment]
The title of this PR is no longer adequately descriptive, can you please update it?

@open-telemetry/collector-approvers please review this PR as it has changed in scope and may have more significant impact than initially proposed.

File: service/config_provider.go
Comment (Aneurysm9) [reply service/config_provider.go]: > // Change configmapprovider.Provider to have a `Scheme()` func on it.

I think I prefer this approach.  We require that providers have a scheme anyways, so why not a way to determine what they think their scheme is?

> should I create a function that converts config maps to []ConfigProviderOption and appends them to provided options to avoid reused code or it's not that big deal?

An unexported helper function to do this might be useful, but at least `service.MustNewConfigProvider()` is expected to go away in the near future, so I'm not sure how big a deal it is.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:88]: +1

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:80]: Same effect. Interface with a private func or a func working on a private type. No real difference. We can discuss this style in a separate issue/PR

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:92]: I am confused about your comment: `I think I would prefer to see these append to the config, allowing for multiple invocations of the options, possibly provided by separate packages.` are you suggesting a global?

Even without the "default" problem, your suggestion does not trivially allow "no converters" case.

Also these append thing is unclear to me, I would rather provide only one "Converter" and a "MultiConverter" implementation if the list is a problem.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go]: > I think I prefer this approach. We require that providers have a scheme anyways, so why not a way to determine what they think their scheme is?

After more thoughts on this, I think is the right thing to do, since this will allow us to use the same pattern in the builder to support more providers - builder will look for NewProvider in the given import (similar to NewFactory for components.

Per @Aneurysm9 comment the `service.MustNewConfigProvider()` is going away, and the other one is a test. I would rather look into an example how a distribution with extra providers will use this. Or an example written for this case.

File: service/config_provider.go
Comment (bogdandrutu) [reply service/config_provider.go:80]: Context: We already have this pattern, which offers the same flexibility/protection. So let's discuss separately.

**Pull request closed without merging**
